### ISSUE TITLE
wal-g releases are incompatible with CentOS 6

### ISSUE THREAD

**Issue by jstaf** [wal-g/wal-g#300]
Because wal-g uses C/C++ extensions (namely the brotli library), it is dependent on the version of glibc where the binary is built. The current build environment makes wal-g incompatible with older OS versions running on many database servers, like CentOS 6 (this is probably true for older Ubuntu and Debian versions, but I haven't tested those).

```
$ ./wal-g 
./wal-g: /lib64/libc.so.6: version `GLIBC_2.14' not found (required by ./wal-g)
$ ldd wal-g
./wal-g: /lib64/libc.so.6: version `GLIBC_2.14' not found (required by ./wal-g)
	linux-vdso.so.1 =>  (0x00007ffd71f10000)
	libm.so.6 => /lib64/libm.so.6 (0x00007f8ba911c000)
	libdl.so.2 => /lib64/libdl.so.2 (0x00007f8ba8f18000)
	liblzo2.so.2 => /usr/lib64/liblzo2.so.2 (0x00007f8ba8cf7000)
	libpthread.so.0 => /lib64/libpthread.so.0 (0x00007f8ba8ada000)
	libc.so.6 => /lib64/libc.so.6 (0x00007f8ba8746000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f8ba93a0000)
```

To resolve this, you should either provide a release where CGo is disabled/brotli compression is totally absent from the binary, or build on an older OS to avoid the glibc versioning issue. For what it's worth, wal-g works on CentOS 6 when it's built on CentOS 6 (I tried this out just now and was able to get a working binary from the latest master).

**Comment by x4m** [comment]
Thanks for reporting this! I will think about how to make releases better...
We can make brotli optional like lzo is now... but then we will have to ship lzo,br,lzo+br, that's not good...
Or we can make releases on CentOS too...

**Comment by jstaf** [comment]
I think the simplest resolution would be to just build releases on something with an older version of glibc. Any older OS should work for this, CentOS 6 is just one example. Let me see if I can dig up what docker image I used to do the build when I get home (created the Dockerfile as a one-off to test things out). 

Regarding lzo+brotli, after second thought I'd almost just include both in every release once the glibc versioning issue gets ironed out. From my understanding of things, brotli gets built into the binary anyways, and lzo is in the standard repositories for most servers, so having lzo/brotli by default isn't a big deal. (Anyone who's migrating to wal-g from wal-e probably already has lzo installed.)

**Comment by SuperQ** [comment]
You should be able to build the binaries statically, this would avoid the cross-distro library issues.

**Comment by Deracination** [comment]
Yes, please provide static builds

**Comment by ernstae** [comment]
I'm really having a frustrating time attempting to build MySQL wal-g binaries for CentOS 7.  Any movement on making binary releases for the additional databases that are supported and other distributions other than Ubuntu?

**Comment by SuperQ** [comment]
The problem is that this codebase uses CGO to include brotli compression. This means that it's impossible to build a real static Go binary.

Without removing this CGO code, this is probably not going to happen.

**Comment by SuperQ** [comment]
You don't need brotli to build, there's makefile pre-build targets that vendor the correct code to build.

But yes, getting rid of brotli by default would make it much easier to publish static builds.

**Comment by jstaf** [comment]
Brotli should definitely stay in the build. Restores using Brotli are about 2-3x faster than with the default LZ4 compression (LZ4 performance is so much worse I think Brotli should actually be made the default compression format). I've only benchmarked this for MySQL, but Postgres and other db performance should be similarly better for Brotli vs. LZ4.

**Comment by SuperQ** [comment]
That doesn't sound correct. brotli is one of the slower decompressors out there today. If you want faster, something like zstd or snappy would be better.

**Comment by SuperQ** [comment]
Another option would be to switch to a pure-go implementation: https://github.com/andybalholm/brotli

**Comment by x4m** [comment]
Brotli can consume slightly more CPU, but we generally don't care as long as the restore is bottlenecked by network or FS. In most cases, brotli is the best codec in WAL-G as of now. It's excluded by default to ease development, but in production it's codec number 1.
ZStd is a great codec but it was corrupting data back in 2018 Zstd was corrupting WAL-G backups https://github.com/DataDog/zstd/issues/39 and we switched to brotli. Maintainers claim to fix this in 2020, so probably we should restore zstd. But personally, I will not switch back from brotli. BTW we had one backup confirmedly corrupted by brotli too. But just one in 4 years and 10^4 clusters. I prefer to think that it was cosmic rays. S3\network corruption is excluded, PGP checksums are OK.
If someone will benchmark pure-go brotli it would be great to switch to it. And forget this glibc nightmare...
BTW we have related #425 and #586.

**Comment by halsbox** [comment]
Hello!
I'd like to suggest the way to make fully portable static builds of wal-g. It does not require to make any changes to source code and works perfectly for me on large variety of linux servers. The key is to statically link everything, including libc and to use musl-libc instead of glibc to minimize the size of binary. I'm using golang:alpine docker image for building and the only change is adding `-linkmode external -extldflags \"-static\"` to ldflags in the build command. E.g. for building main/pg I use:
```
go build -mod vendor -tags "$(BUILD_TAGS)" -o wal-g --ldflags "-s -w -linkmode external -extldflags \"-static\" -X github.com/wal-g/wal-g/cmd/pg.buildDate=`date -u +%Y.%m.%d_%H:%M:
%S` -X github.com/wal-g/wal-g/cmd/pg.gitRevision=`git rev-parse --short HEAD` -X github.com/wal-g/wal-g/cmd/pg.walgVersion=`git tag -l --points-at HEAD`"
```
instead of 
```
go build -mod vendor -tags "$(BUILD_TAGS)" -o wal-g -ldflags "-s -w -X github.com/wal-g/wal-g/cmd/pg.buildDate=`date -u +%Y.%m.%d_%H:%M:%S` -X github.com/wal-g/wal-g/cmd/pg.gitRevision=`git rev-parse --short HEAD` -X github.com/wal-g/wal-g/cmd/pg.walgVersion=`git tag -l --points-at HEAD`"
```
For wal-g-pg the result binary is 41.6M in size (official non static wal-g-gp-ubuntu-20.04-amd64 is 40.6 MB) and can be reduced down to 12.2M by compressing using upx. Resulting binary is tested to work perfectly even on very old linux distributions.

**Comment by jstaf** [comment]
I tested out a variation on static build instructions and the static build does work as described though there are some interesting NSS errors with non-UNIX user accounts. Running any `wal-g` command on user accounts managed by services like FreeIPA/sssd or GCP's OSLogin will result in errors like this:

```
$ wal-g backup-list
ERROR: 2022/07/14 17:16:17.325247 user: unknown userid 1092191234
```

I think this is more a musl libc limitation than a problem with the static build though. And you'd never really be using wal-g with non-UNIX users anyways, you'd usually be using it after sudo-ing to another "valid" user.

**Comment by halsbox** [comment]
You are right, it's because wal-g tries to find it's own config file in current user's home and musl-libs does know noting about nss and just looks in /etc/passwd and errors on not finding current user id there. The workaround is to provide wal-g with a path to config file using --config flag, this way you won't get any "unknown userid" error even with nss-managed user.

**Comment by jstaf** [comment]
I put together a PR that makes it easier to create static wal-g builds here: https://github.com/wal-g/wal-g/pull/1315
